rcl/test/rcl/test_publisher.cpp

```diff
@@ -490,3 +535,294 @@ TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_invalid_publish
+  test_msgs__msg__BasicTypes msg;
+  test_msgs__msg__BasicTypes__init(&msg);
+  msg.int64_value = 42;
+  ret = rcl_publish(&publisher, &msg, nullptr);
+  test_msgs__msg__BasicTypes__fini(&msg);
+  EXPECT_EQ(RCL_RET_ERROR, ret) << rcl_get_error_string().str;
+  EXPECT_TRUE(rcl_error_is_set());
+  rcl_reset_error();
+}
+
+// Mocking rmw_publish_serialized_message to make rcl_publish_serialized_message fail
+TEST_F(
+  CLASSNAME(TestPublisherFixtureInit, RMW_IMPLEMENTATION), test_mock_publish_serialized_message)
+{
+  rcl_serialized_message_t serialized_msg = rmw_get_zero_initialized_serialized_message();
+  size_t initial_capacity_ser = 0u;
+  rcl_allocator_t allocator = rcl_get_default_allocator();
+  ASSERT_EQ(
+    RCL_RET_OK, rmw_serialized_message_init(
+      &serialized_msg, initial_capacity_ser, &allocator)) << rcl_get_error_string().str;
+  const char * test_string = "testing";
+  test_msgs__msg__Strings msg;
+  test_msgs__msg__Strings__init(&msg);
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    test_msgs__msg__Strings__fini(&msg);
+  });
+  ASSERT_TRUE(rosidl_runtime_c__String__assign(&msg.string_value, test_string));
+  ASSERT_STREQ(msg.string_value.data, test_string);
+  ret = rmw_serialize(&msg, ts, &serialized_msg);
+  ASSERT_EQ(RMW_RET_OK, ret);
+
+  {
+    // Test normal usage of the function rcl_publish_serialized_message
+    // returning unexpected RMW_RET_ERROR
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_publish_serialized_message, [](auto...) {return RMW_RET_ERROR;});
+    ret = rcl_publish_serialized_message(&publisher, &serialized_msg, nullptr);
+    EXPECT_EQ(RCL_RET_ERROR, ret) << rcl_get_error_string().str;
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+  {
+    // Repeat, but now returning BAD_ALLOC
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rmw_publish_serialized_message, [](auto...) {return RMW_RET_BAD_ALLOC;});
+    ret = rcl_publish_serialized_message(&publisher, &serialized_msg, nullptr);
+    EXPECT_EQ(RCL_RET_BAD_ALLOC, ret) << rcl_get_error_string().str;
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+}
+
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, ==)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, <)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, >)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, !=)
+
+TEST_F(CLASSNAME(TestPublisherFixture, RMW_IMPLEMENTATION), test_mock_rcutils_string_map_init) {
+  rcl_publisher_t publisher = rcl_get_zero_initialized_publisher();
+  const rosidl_message_type_support_t * ts =
+    ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, Strings);
+  const char * topic_name = "chatter";
+  rcl_publisher_options_t publisher_options = rcl_publisher_get_default_options();
+  rcl_ret_t ret = RCL_RET_OK;
+
+  auto mock = mocking_utils::patch(
+    "lib:rcl", rcutils_string_map_init, [](auto...) {return RCUTILS_RET_ERROR;});
```

Comment:
@Blast545 meta: if many of your tests simply force a return value, consider adding a convenience macro to `patch.hpp`, like:

```c++
#define patch_and_return(scope, function, return_value) patch(scope, function, [&](auto && ...) {return return_value;})
```

Name is up for debate :)

Reply:
Good idea, I'll add that macro to `patch.hpp`

Reply:
See d4228e4